packages-exp/auth-exp/src/platform_browser/messagechannel/sender.ts

```diff
@@ -0,0 +1,143 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  SenderRequest,
+  EventType,
+  ReceiverMessageEvent,
+  MessageError,
+  SenderMessageEvent,
+  Status,
+  ReceiverMessageResponse,
+  ReceiverResponse,
+  TimeoutDuration
+} from './index';
+
+interface MessageHandler {
+  messageChannel: MessageChannel;
+  onMessage: EventListenerOrEventListenerObject;
+}
+
+/**
+ * Interface for sending messages and waiting for a completion response.
+ *
+ * @internal
+ */
+export class Sender {
+  private readonly handlers: Set<MessageHandler> = new Set();
+
+  constructor(private readonly target: ServiceWorker) {}
+
+  /**
+   * Unsubscribe the handler and remove it from our tracking Set.
+   *
+   * @param handler - The handler to unsubscribe.
+   */
+  private removeMessageHandler(handler: MessageHandler): void {
+    if (handler.messageChannel) {
+      handler.messageChannel.port1.removeEventListener(
+        'message',
+        handler.onMessage
+      );
+      handler.messageChannel.port1.close();
+    }
+    this.handlers.delete(handler);
+  }
+
+  private generateEventId(prefix: string = '', digits: number = 20): string {
```

Comment:
nit: make this a free-floating function since it doesn't use `this`

Reply:
done